apps/emqx/src/emqx_shared_sub.erl

```diff
@@ -239,6 +240,9 @@ with_redispatch_to(#message{qos = ?QOS_0} = Msg, _Group, _Topic) ->
 with_redispatch_to(Msg, Group, Topic) ->
     emqx_message:set_headers(#{redispatch_to => ?REDISPATCH_TO(Group, Topic)}, Msg).
 
+with_shared_record(Msg, Group, Topic) ->
+    emqx_message:set_headers(#{shared_record => emqx_topic:make_shared_record(Group, Topic)}, Msg).
```

Comment:
I think we only need one of such headers.
they serve different purposes, but it's exactly the same information.

Reply:
the only difference is QoS0, which can be handled in `get_redispatch_to` function instead.

Reply:
to support rolling upgrade, my suggestion is to use ?REDISPATCH_TO instead of adding a new header.